Automattic/woocommerce-services
# Ensure rates prop exists before searching it

To test:

1. Follow the steps on this server PR to get a rates response with an error:
https://github.com/Automattic/woocommerce-connect-server/pull/802

Or, use a fake response (listed at the end of this description) on the server.

2. Open the printing flow to get a rates request

3. See that there is no error in the console (without this fix, you'd see `Uncaught (in promise) TypeError: Cannot read property 'find' of undefined`)

4. See this for rates:

<img width="564" alt="rates-with-errors" src="https://cloud.githubusercontent.com/assets/11487924/23436341/2913aa0a-fdd8-11e6-92e5-8f6f7bee3ec2.png">

And here's what happens when clicking on the drop-down:

<img width="561" alt="rates-with-errors-no-selections-available" src="https://cloud.githubusercontent.com/assets/11487924/23436351/2fb59c06-fdd8-11e6-9094-6759bada2f5c.png">

# Communicating unavailable rates

I think this needs a bit of design help (not necessarily a blocker though), what do you @kellychoffman? At the very least, I was thinking of removing the drop-downs and explaining that there is an error:

<img width="570" alt="screen shot 2017-02-28 at 5 11 54 pm" src="https://cloud.githubusercontent.com/assets/11487924/23436573/0439100c-fdd9-11e6-813e-a967147fdc73.png">

# Using a fake response

Above line 169 in lib/shipping/handler.js on the server, add this fake response code:

```
const mockUspsRates = {
  shipment_id: 'rates_weight_0_individual',
  rates: [ {
    title: 'USPS - Priority Mail',
    rate_id: 'rate_123',
    carrier_id: 'usps',
    rate: 2.87,
    service_id: 'priority',
  } ],
};
const mockFedexError = {
  code: 'ERR2',
  message: 'There was a fedex error!',
};
const fakeResponse = {
  rates: {
    weight_0_individual: {
      shipment_id: 'abcdef',
      rates: [ Object.assign( {}, mockUspsRates.rates[ 0 ], { title: 'USPS - Priority Mail translated' } ) ],
      errors: [ Object.assign( { carrier: 'fedex' }, mockFedexError ) ],
    },
    weight_1_individual: {
      shipment_id: 'abcdef',
      rates: [ Object.assign( {}, mockUspsRates.rates[ 0 ], { title: 'USPS - Priority Mail translated' } ) ],
      errors: [ Object.assign( { carrier: 'fedex' }, mockFedexError ) ],
    },
  },
};
const fakeResponse2 = {
  rates: {
    weight_0_individual: {
      errors: [ Object.assign( { carrier: 'fedex' }, mockFedexError ) ],
    },
    weight_1_individual: {
      errors: [ Object.assign( { carrier: 'fedex' }, mockFedexError ) ],
    },
  },
};
```

Then, on line  169, use `reply( fakeResponse );` You may need to change the package IDs.
